examples/shor/functions.py

```diff
@@ -0,0 +1,399 @@
+    modular exponentiation. |1>|x>|0> ---> |1>|(a*x)%N>|0>
+    Args:
+        c (int): qubit acting as control.
+        x (list): quantum register encoding the number of times a is multiplied.
+        b (list): quantum register initialized at |0>
+        a (int): number to multiply.
+        N (int): modulo.
+        ancilla (int): extra quantum register needed for the modular addition.
+        n (int): number of bits needed to encode N
+
+    Returns:
+        generator with the required quantum gates applied on the quantum circuit.
+    """
+    yield c_mult_mod(c, x, b, a, N, ancilla, n)
+    for i in range(1, len(b)):
+        yield gates.SWAP(b[i], x[i-1]).controlled_by(c)
+    a_inv = mod_inv(a, N)
+    yield i_c_mult_mod(c, x, b, a_inv, N, ancilla, n)
+
+
+def find_factor_of_prime_power(N):
+    """Classical algorithm to check if given number is a prime power.
+    Args:
+        N (int): number to factorize.
+
+    Returns:
+        If the number is not a prime power: None
+        If the number is a prime power: f1 (int) or f2 (int)
+    """
+    for i in range(2, int(np.floor(np.log2(N))) + 1):
+        f = N**(1/i)
+        f1 = np.floor(f)
+        if f1**i == N:
+            return f1
+        f2 = np.ceil(f)
+        if f2**i == N:
+            return f2
+    return None
+
+
+def egcd(a, b):
+    """Extended Euclid's algorithm for the modular inverse calculation.
+    """
+    if a == 0:
+        return (b, 0, 1)
+    else:
+        g, y, x = egcd(b % a, a)
+        return (g, x - (b // a) * y, y)
+
+
+def mod_inv(a, N):
+    """Classical algorithm to calculate the inverse of a modulo N.
+    Args:
+        a (int): number to invert.
+        N (int): modulo.
+
+    Returns:
+        x%N (int): inverse of a modulo N.
+    """
+    g, x, y = egcd(a, N)
+    if g != 1:
+        raise ValueError('modular inverse does not exist')
+    else:
+        return x % N
+
+
+def quantum_order_finding_full(N, a):
+    """Quantum circuit that performs the order finding algorithm using a fully quantum iQFT.
+    Args:
+        N (int): number to factorize.
+        a (int): chosen number to use in the algorithm.
+
+    Returns:
+        s (float): value of the state measured by the quantum computer.
+    """
+    print('  - Performing algorithm using a fully quantum iQFT.\n')
+    # Creating the parts of the needed quantum circuit.
+    n = int(np.ceil(np.log2(N)))
+    b = [i for i in range(n+1)]
+    x = [n+1+i for i in range(n)]
+    ancilla = 2*n + 1
+    q_reg = [2*n + 2 + i for i in range(2*n)]
+    circuit = Circuit(4*n+2)
+    print(f'  - Total number of qubits used: {4*n+2}.\n')
+
+    # Building the quantum circuit
+    for i in range(len(q_reg)):
+        circuit.add(gates.H(q_reg[i]))
+    circuit.add(gates.X(x[len(x)-1]))
+    for i in (range(len(q_reg))):
+        a_i = (a**(2**(i)))
+        circuit.add(c_U(q_reg[i], x, b, a_i, N, ancilla, n))
+    circuit.add(i_qft(q_reg))
```

Comment:
Small simplifications:
```Python
circuit.add((gates.H(q) for q in q_reg))
circuit.add(gates.X(x[-1]))
circuit.add((c_U(q, x, b, a**(2**i), N, ancilla, n) for i, q in enumerate(q_reg)))
circuit.add(i_qft(q_reg))
```
Similar things can probably be applied in other places but it is minor/optional, just keep whatever looks best for you.